### PR TITLE
Implement `engine_exchangeCapabilities` method

### PR DESCRIPTION
**Pull request by rubo**
## Changes

- Implemented `engine_exchangeCapabilities` method as specified in [the spec](https://github.com/ethereum/execution-apis/blob/main/src/engine/common.md#engine_exchangecapabilities)
- In addition, the execution client warns if it has more methods (considers only the activated ones on its end) than found in the consensus client.

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Merge.Plugin/EngineRpcModule.cs [hunk 1]
```diff
@@ -46,6 +49,16 @@ public partial class EngineRpcModule : IEngineRpcModule
 
     public ResultWrapper<ExecutionStatusResult> engine_executionStatus() => _executionStatusHandler.Handle();
 
+    public ResultWrapper<IEnumerable<string>> engine_getCapabilities()
+    {
+        _capabilities ??= typeof(IEngineRpcModule).GetMethods()
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Merge.Plugin/EngineRpcModule.cs:54]: no caching

**Review by MarekM25 — CHANGES_REQUESTED** [review]
no caching

**Comment by rubo** [comment]
> no caching

Take a closer look.

File: src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.V1.cs [hunk 2]
```diff
@@ -1661,6 +1661,22 @@ public async Task payloadV1_latest_block_after_reorg()
         }
     }
 
+    [Test]
+    public async Task Should_return_capabilities()
+    {
+        using var chain = await CreateBlockChain();
+        var rpcModule = CreateEngineModule(chain);
+
+        var expected = typeof(IEngineRpcModule).GetMethods()
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.V1.cs:1670]: your tests are the same as the code, so we're not testing anything IMO

File: src/Nethermind/Nethermind.Merge.Plugin/IEngineRpcModule.cs [hunk 3]
```diff
@@ -19,6 +20,12 @@ public partial interface IEngineRpcModule : IRpcModule
         IsImplemented = true)]
     ResultWrapper<ExecutionStatusResult> engine_executionStatus();
 
+    [JsonRpcMethod(
+        Description = "Returns the currently supported list of Engine API methods.",
+        IsSharable = true,
+        IsImplemented = true)]
+    ResultWrapper<IEnumerable<string>> engine_getCapabilities();
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Merge.Plugin/IEngineRpcModule.cs:27]: let's confirm the method name

**Review by MarekM25 — CHANGES_REQUESTED** [review]
No caching methods name

**Review by MarekM25 — APPROVED** [review]

File: src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.V1.cs (hunk 2 above)
Comment (rubo) [reply src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.V1.cs:1670]: How's that? If you change something in the method test will catch it. The fact they are the same doesn't mean they are connected or affect each other.

**Review by MarekM25 — CHANGES_REQUESTED** [review]
We can also print warnings when our exchange capabilities are not the same.

File: src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.V1.cs (hunk 2 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.V1.cs:1670]: you may introduce the same bug in both, test should hardcode your expectations and should read like a spec

**Comment by rubo** [comment]
> We can also print warnings when our exchange capabilities are not the same.

It should work a bit differently. A warning is needed only when we miss a capability in the request list. Otherwise, no warning. For instance, if we have more capabilities.

File: src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.V1.cs [hunk 4]
```diff
@@ -1661,6 +1663,46 @@ public async Task payloadV1_latest_block_after_reorg()
...
 
+    [Test]
+    public async Task Should_return_capabilities()
+    {
+        using var chain = await CreateBlockChain();
+        var rpcModule = CreateEngineModule(chain);
+        var expected = new[]
+        {
+            "engine_exchangeTransitionConfigurationV1",
+            "engine_executionStatus",
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.V1.cs:1674]: And the value of these tests has been already proved
"engine_executionStatus"
This method can be removed

Comment (MarekM25) [reply src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.V1.cs:1674]: and even it should be removed from code

The method is not needed now

File: src/Nethermind/Nethermind.Merge.Plugin/EngineRpcModule.cs [hunk 5]
```diff
@@ -44,6 +48,47 @@ public partial class EngineRpcModule : IEngineRpcModule
         _logger = logManager.GetClassLogger();
     }
 
+    public async Task<ResultWrapper<IEnumerable<string>>> engine_exchangeCapabilities(IEnumerable<string> methods)
+    {
+        ArgumentNullException.ThrowIfNull(methods);
+
+        if (await _locker.WaitAsync(1_000))
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Merge.Plugin/EngineRpcModule.cs:55]: _locker is not needed here. Please remove

File: src/Nethermind/Nethermind.Merge.Plugin/EngineRpcModule.cs [hunk 6]
```diff
@@ -44,6 +48,47 @@ public partial class EngineRpcModule : IEngineRpcModule
...
+                _capabilities ??= typeof(IEngineRpcModule).GetMethods()
+                    .Select(m => m.Name)
+                    .Where(m => !m.Equals(nameof(engine_exchangeCapabilities), StringComparison.Ordinal))
+                    .Order();
+
+                var unsupported = methods.Except(_capabilities);
+
+                if (unsupported.Any())
+                {
+                    if (_logger.IsWarn) _logger.Warn($"Unsupported capabilities: {string.Join(", ", unsupported)}");
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Merge.Plugin/EngineRpcModule.cs:70]: I think we should create a separate handler for the request.

File: src/Nethermind/Nethermind.Core.Test/Blockchain/TestBlockchain.cs [hunk 7]
```diff
@@ -31,399 +30,395 @@
...
-        public ITxPool TxPool { get; set; } = null!;
-        public IDb CodeDb => DbProvider.CodeDb;
-        public IBlockProcessor BlockProcessor { get; set; } = null!;
-        public IBlockchainProcessor BlockchainProcessor { get; set; } = null!;
-
-        public IBlockPreprocessorStep BlockPreprocessorStep { get; set; } = null!;
-
-        public IBlockProcessingQueue BlockProcessingQueue { get; set; } = null!;
-        public IBlockTree BlockTree { get; set; } = null!;
-
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Core.Test/Blockchain/TestBlockchain.cs:62]: What was changed in TestBlockchain?

**Review by MarekM25 — CHANGES_REQUESTED** [review]
still a few things to address

File: src/Nethermind/Nethermind.Core.Test/Blockchain/TestBlockchain.cs (hunk 7 above)
Comment (rubo) [reply src/Nethermind/Nethermind.Core.Test/Blockchain/TestBlockchain.cs:62]: Added mutator for `LogManager`

File: src/Nethermind/Nethermind.Merge.Plugin/EngineRpcModule.cs (hunk 6 above)
Comment (rubo) [reply src/Nethermind/Nethermind.Merge.Plugin/EngineRpcModule.cs:70]: I don't see why. Could you please elaborate what's the benefit of this complication?

File: src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.V1.cs (hunk 4 above)
Comment (rubo) [reply src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.V1.cs:1674]: Alright, not in this PR though.

File: src/Nethermind/Nethermind.Merge.Plugin/Handlers/ExchangeCapabilitiesHandler.cs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        _logger = logManager.GetClassLogger();
+
+        if (_capabilities.Count == 0)
+        {
+            var spec = specProvider.GetSpec((long.MaxValue, ulong.MaxValue));
+
+            _capabilities[nameof(IEngineRpcModule.engine_exchangeTransitionConfigurationV1)] = true;
+            _capabilities[nameof(IEngineRpcModule.engine_executionStatus)] = true;
+            _capabilities[nameof(IEngineRpcModule.engine_forkchoiceUpdatedV1)] = true;
+            _capabilities[nameof(IEngineRpcModule.engine_forkchoiceUpdatedV2)] = spec.WithdrawalsEnabled;
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Merge.Plugin/Handlers/ExchangeCapabilitiesHandler.cs:35]: don't you think that some list of the methods will be better?

**Review by MarekM25 — APPROVED** [review]

**Pull request merged**